File: src/lib/platform.ts
import type { PlatformDownloads } from "./types";

export type { PlatformDownloads };

export type PlatformKey = keyof PlatformDownloads;

export function emptyPlatforms(): PlatformDownloads {
  return { macos: 0, windows: 0, linux: 0, other: 0 };
}

export function classifyAsset(name: string): PlatformKey {
  const n = name.toLowerCase();
  if (
    n.includes("windows") ||
    n.includes("win32") ||
    n.includes("win64") ||
    n.includes("win-") ||
    n.includes("-win.") ||
    n.includes("-pc-windows") ||
    n.endsWith(".msi") ||
    n.endsWith(".msix") ||
    n.endsWith(".exe") ||
    n.endsWith(".nupkg")
  ) {
    return "windows";
  }
  if (
    n.includes("macos") ||
    n.includes("darwin") ||
    n.includes("osx") ||
    n.includes("apple-darwin") ||
    n.endsWith(".dmg") ||
    n.endsWith(".pkg") ||
    n.endsWith(".app")
  ) {
    return "macos";
  }
  if (
    n.includes("linux") ||
    n.includes("appimage") ||
    n.includes("unknown-linux") ||
    n.endsWith(".deb") ||
    n.endsWith(".rpm") ||
    n.endsWith(".appimage")
  ) {
    return "linux";
  }
  return "other";
}

export function addAsset(platforms: PlatformDownloads, name: string, count: number) {
  platforms[classifyAsset(name)] += count;
}

export function platformsFromAssets(
  assets: { name: string; downloadCount: number }[],
): PlatformDownloads {
  const platforms = emptyPlatforms();
  for (const asset of assets) addAsset(platforms, asset.name, asset.downloadCount);
  return platforms;
}

export function sumPlatforms(list: Array<PlatformDownloads | undefined | null>): PlatformDownloads {
  const platforms = emptyPlatforms();
  for (const item of list) {
    if (!item) continue;
    platforms.macos += item.macos;
    platforms.windows += item.windows;
    platforms.linux += item.linux;
    platforms.other += item.other;
  }
  return platforms;
}

export function platformItems(platforms: PlatformDownloads) {
  return (
    [
      { label: "macOS", value: platforms.macos },
      { label: "Windows", value: platforms.windows },
      { label: "Linux", value: platforms.linux },
      { label: "Other", value: platforms.other },
    ] as const
  )
    .filter((item) => item.value > 0)
    .sort((a, b) => b.value - a.value)
    .map((item) => ({ label: item.label, value: item.value }));
}

export function shortPath(path: string, fullName: string) {
  const prefix = `/${fullName}`;
  if (path === prefix || path === prefix + "/") return "Overview";
  if (path.startsWith(prefix)) {
    const rest = path.slice(prefix.length);
    return rest.startsWith("/") ? rest : `/${rest}`;
  }
  return path;
}

File: src-tauri/src/platform.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Platform {
    Macos,
    Windows,
    Linux,
    Other,
}

pub fn classify_asset(name: &str) -> Platform {
    let n = name.to_ascii_lowercase();
    if is_windows(&n) {
        Platform::Windows
    } else if is_macos(&n) {
        Platform::Macos
    } else if is_linux(&n) {
        Platform::Linux
    } else {
        Platform::Other
    }
}

fn is_windows(n: &str) -> bool {
    n.contains("windows")
        || n.contains("win32")
        || n.contains("win64")
        || n.contains("win-")
        || n.contains("-win.")
        || n.contains("-pc-windows")
        || n.ends_with(".msi")
        || n.ends_with(".msix")
        || n.ends_with(".exe")
        || n.ends_with(".nupkg")
}

fn is_macos(n: &str) -> bool {
    n.contains("macos")
        || n.contains("darwin")
        || n.contains("osx")
        || n.contains("apple-darwin")
        || n.ends_with(".dmg")
        || n.ends_with(".pkg")
        || n.ends_with(".app")
}

fn is_linux(n: &str) -> bool {
    n.contains("linux")
        || n.contains("appimage")
        || n.contains("unknown-linux")
        || n.ends_with(".deb")
        || n.ends_with(".rpm")
        || n.ends_with(".appimage")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn classifies_common_release_assets() {
        assert_eq!(classify_asset("asterism_0.1.0_aarch64.dmg"), Platform::Macos);
        assert_eq!(classify_asset("app-macos-arm64.pkg"), Platform::Macos);
        assert_eq!(classify_asset("app-0.3.0-windows-x64.msi"), Platform::Windows);
        assert_eq!(classify_asset("app_0.1.0_x64-setup.exe"), Platform::Windows);
        assert_eq!(classify_asset("app-0.3.0-linux-amd64.AppImage"), Platform::Linux);
        assert_eq!(classify_asset("app_0.1.0_amd64.deb"), Platform::Linux);
        assert_eq!(classify_asset("source.tar.gz"), Platform::Other);
    }
}
